### PR TITLE
add `setup.py`

### PR DESCRIPTION
**Pull request by ghost**
`windows-container` is a pretty good. It is better for it to appear on [PyPI](https://pypi.org) as a python package for everyone to download and install. For that, I wrote `setup.py`, it

* makes building easier. To use Visual Studio solution file is not considered so good because it is tied with version. For example, to build in Visual Studio 2017, the toolchain config needs to upgrade. the `setup.py` provides universal build way `python setup.py build`

* makes it be able to upload to PyPI. After this everyone could use `pip install windows-container` to install windows-container

Note that the `setup.py` is not complete. Some info of package like `author`, `version` etc is not there. You might need to complete it

I'd like to help with work on uploading to PyPI. Also you could entrust to me

I have also written some script to run building on [appveyor](https://www.appveyor.com)

### CODE REVIEW

**Comment by iceboy233** [comment]
Thanks for the contribution! This is something I wanted to do but I didn't have the chance/need to do it.

Have you tested with your setup.py? What compiler version does it typically use? I think python by default uses a lower version of C++ compiler (especially for python 2.x), and we need to migrate off some usages of new C++ features to make it work.

Also let's not make `winc` python-centric (e.g. other users can use native C++ interface, or create bindings for other languages). Having a `setup.py` in root directory is fine though.

**Comment by ghost** [comment]
>  I think python by default uses a lower version of C++ compiler

It is a fact that it can only work with python3.5+ because the `distutils` of lower versions of python are always trying to build with MSVC9.0 that makes build fail. I'm also trying to find a solution

> Also let's not make winc python-centric (e.g. other users can use native C++ interface, or create bindings for other languages). Having a setup.py in root directory is fine though.

But another problem is that vs project file is tied with the version of VS's that means one could have problems to build on VS2017 like the need of upgrading toolchain config and Windows SDK. And solution file is generated by machine so that I consider it as outcome not the source which considered not so good to appear in a repo. Besides, it's not so clear to read by human and port. Did you consider `cmake` or others?

**Comment by iceboy233** [comment]
We can make this work under `vc9`.

**Comment by ghost** [comment]
It seems that vc9 could not compile because of new C++ features. I think it's better way to do some hack on `distutils`

**Comment by iceboy233** [comment]
> It seems that vc9 could not compile because of new C++ features. I think it's better way to do some hack on `distutils`

No. It's important to use same compiler version on Windows to avoid DLL hell. In this case we should use the same compiler version as Python binary.

btw are you creating any kind of online judge service? I'm interested in that :)

**Comment by ghost** [comment]
> No. It's important to use same compiler version on Windows to avoid DLL hell. In this case we should use the same compiler version as Python binary.

That sounds not so good

**Comment by iceboy233** [comment]
Could you explain why it's not so good? See also https://wiki.python.org/moin/WindowsCompilers .

**Comment by ghost** [comment]
...I mean that is not a good message for me...

**Comment by ghost** [comment]
I found a workable way to be built with py2.7

You can have a look at [yawinpty](https://pypi.python.org/pypi/yawinpty). That's a package written by me wrappers [winpty](https://github.com/rprichard/winpty). winpty requires new version of VS to build and fail with MSVC9.0. My solution is to first compile winpty as DLL with VS2015, export symbols and let my part of code compiled by MSVC9.0 dynamically link to `winpty.dll`

But there's still problem. The steps of building seem to be difficult to write to `setup.py`

File: .gitignore [hunk 1]
```diff
@@ -10,3 +10,24 @@
 Debug/
 Release/
 x64/
+
+# Python Distribution / packaging
```
Comment (iceboy233) [inline .gitignore:14]: Could you limit this only to the actual generated files during a python build?

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+                # see https://msdn.microsoft.com/en-us/library/windows/desktop/ms683219(v=vs.85).aspx
+                ('PSAPI_VERSION', '1'),
+                # FIXME: _HAS_EXCEPTIONS only has effect when /MT
+                # But python builds it with /MD
+                # ('_HAS_EXCEPTIONS', '0'),
+                # For future?
+                ('BINDING_PYTHON_EXPORTS', None),
+                # Use UNICODE
+                ('_UNICODE', None),
+                ('UNICODE', None)],
```
Comment (iceboy233) [inline setup.py]: Could you reformat this to make right brackets to have their own line?

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            include_dirs = [
+                'include',
+                '.'],
+            libraries = [
+                'Advapi32',
```
Comment (iceboy233) [inline setup.py]: Is uppercase required? If not, use lowercase.

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+                'Ntdll',
+                'User32',
+                'Psapi'],
+            sources = glob(path.join('core', '*.cc')) +
```
Comment (iceboy233) [inline setup.py]: You can write `core/*.cc`, it's not a externally controlled variable.

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            define_macros = [
+                # see https://msdn.microsoft.com/en-us/library/windows/desktop/ms683219(v=vs.85).aspx
+                ('PSAPI_VERSION', '1'),
+                # FIXME: _HAS_EXCEPTIONS only has effect when /MT
+                # But python builds it with /MD
+                # ('_HAS_EXCEPTIONS', '0'),
+                # For future?
+                ('BINDING_PYTHON_EXPORTS', None),
+                # Use UNICODE
+                ('_UNICODE', None),
```
Comment (iceboy233) [inline setup.py]: I think this is not needed. The code uses the XxxW form to call API, bypassing the unicode macro.

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    ext_modules = [
+        Extension('winc',
+            define_macros = [
+                # see https://msdn.microsoft.com/en-us/library/windows/desktop/ms683219(v=vs.85).aspx
+                ('PSAPI_VERSION', '1'),
+                # FIXME: _HAS_EXCEPTIONS only has effect when /MT
+                # But python builds it with /MD
+                # ('_HAS_EXCEPTIONS', '0'),
+                # For future?
+                ('BINDING_PYTHON_EXPORTS', None),
```
Comment (iceboy233) [inline setup.py]: What's this?

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+from glob import glob
+
+setup(
+    name = 'windows-container',
+    ext_modules = [
+        Extension('winc',
+            define_macros = [
+                # see https://msdn.microsoft.com/en-us/library/windows/desktop/ms683219(v=vs.85).aspx
+                ('PSAPI_VERSION', '1'),
+                # FIXME: _HAS_EXCEPTIONS only has effect when /MT
```
Comment (iceboy233) [inline setup.py:16]: fyi https://stackoverflow.com/questions/553103/can-i-disable-exceptions-in-stl

File: .appveyor.ps1 [hunk 8]
```diff
@@ -0,0 +1,18 @@
+$ErrorActionPreference = "Stop"
+Get-ChildItem C:\Python* | ForEach-Object{
+    $python = $_.Name
+    if($python -in 'Python35', 'Python35-x64', 'Python36', 'Python36-x64'){
```
Comment (iceboy233) [inline .appveyor.ps1]: space between `if` and `(`, also between `)` and `{`

**Review by iceboy233 — CHANGES_REQUESTED** [review]
ok. Let's proceed with only python 3.5+ support. We can add support to python 3.4- in the future if needed.  Just keep in mind that we are not doing any kind of hack to `distutils` for toolchain versions.

Could you also update README.md to reflect the changes?

File: setup.py (hunk 7 above)
Comment (ghost) [reply setup.py:16]: no affect, I have tried before

it makes build fail

```
libcpmt.lib(xthrow.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in container.obj
```

**Comment by ghost** [comment]
What you need to do next:

1. complete fields of `setup.py` like `author`, `version` etc
2. add `MANIFEST.in` to point out source files for `setuptools`
3. add this repo to [appveyor](https://www.appveyor.com) to auto build for wheels
  default
4. upload wheels and sources tarballs to PyPI using [twine](https://pypi.org/project/twine)

**Review by iceboy233 — APPROVED** [review]

**Comment by iceboy233** [comment]
Looks mostly good. Feel free to make final changes and/or merge this pull request. Thanks again.

**Comment by ghost** [comment]
ok. I think no more change is needed

I'm forwarding to see it appears on PyPI in the future

**Pull request merged**